File: apps/api/app/models/pipeline.py
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field

NodeKind = Literal["agent", "api", "service", "trigger", "end"]
WireKind = Literal["a2a", "x402", "algo_transfer"]
LogLevel = Literal["info", "success", "warning", "error"]


class Position(BaseModel):
    x: float
    y: float


class NodeData(BaseModel):
    label: str
    role: Optional[str] = None
    description: Optional[str] = None
    systemPrompt: Optional[str] = None
    enabledTools: List[str] = Field(default_factory=list)
    priceAlgo: float = 0
    serviceUrl: Optional[str] = None
    serviceKind: Optional[Literal["weather", "search", "custom"]] = None
    upstreamX402: bool = False
    treasuryAddress: Optional[str] = None


class PipelineNode(BaseModel):
    id: str
    type: NodeKind
    position: Position
    data: NodeData


class PipelineEdge(BaseModel):
    id: str
    source: str
    target: str
    wireType: WireKind
    label: Optional[str] = None


class DeployPipelineRequest(BaseModel):
    name: str
    network: Literal["algorand-testnet"] = "algorand-testnet"
    nodes: List[PipelineNode]
    edges: List[PipelineEdge]


class RuntimeLog(BaseModel):
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    level: LogLevel
    message: str
    nodeId: Optional[str] = None
    eventType: Optional[Literal["start", "progress", "output", "done", "error"]] = None
    output: Optional[str] = None
    txId: Optional[str] = None


class DeployedNode(BaseModel):
    id: str
    type: NodeKind
    walletAddress: Optional[str] = None
    balanceAlgo: Optional[float] = None
    explorerUrl: Optional[str] = None


class DeployPipelineResponse(BaseModel):
    pipelineId: str
    status: Literal["live"] = "live"
    endpoint: str
    priceAlgo: float
    network: str
    paymentWallet: Optional[str] = None
    loraUrl: Optional[str] = None
    nodes: List[DeployedNode]
    logs: List[RuntimeLog]


class WalletBalance(BaseModel):
    nodeId: str
    address: str
    balanceAlgo: float
    explorerUrl: Optional[str] = None


class BalanceResponse(BaseModel):
    pipelineId: str
    balances: List[WalletBalance]


class FundIntentResponse(BaseModel):
    pipelineId: str
    nodeId: str
    address: str
    faucetUrl: str
    loraUrl: Optional[str] = None
    qrValue: str


class RunPipelineRequest(BaseModel):
    query: Optional[str] = None
    payload: Dict[str, Any] = Field(default_factory=dict)


class RunPipelineResponse(BaseModel):
    runId: str
    pipelineId: str
    result: str
    settlementMode: Literal["demo", "payment_response"]
    logs: List[RuntimeLog]


class PaymentRequiredResponse(BaseModel):
    pipelineId: str
    amountAlgo: float
    wallet: str
    network: str
    facilitator: str
    endpoint: str
    message: str


class InternalToolInvokeResponse(BaseModel):
    tool: str
    title: str
    summary: str
    raw: Dict[str, Any] = Field(default_factory=dict)

File: apps/web/src/types/pipeline.ts
import type { Edge, Node } from "@xyflow/react";

export type NodeKind = "agent" | "api" | "service" | "trigger" | "end";
export type WireKind = "a2a" | "x402" | "algo_transfer" | "connection";
export type LogLevel = "info" | "success" | "warning" | "error";

export interface PipelineNodeData extends Record<string, unknown> {
  kind: NodeKind;
  label: string;
  requestMethod?: "GET" | "POST" | "PUT";
  testRequestBody?: string;
  role?: string;
  description?: string;
  systemPrompt?: string;
  enabledTools?: string[];
  priceAlgo?: number;
  serviceUrl?: string;
  serviceKind?: "weather" | "search" | "custom" | "gmail" | "crypto" | "chart" | "risk"
    | "gemini" | "openai" | "claude" | "mistral";
  triggerKind?: "webhook" | "manual" | "schedule" | "chat";
  apiKey?: string;            // AI Model API key — stored client-side only
  upstreamX402?: boolean;
  treasuryAddress?: string;
  gmailTo?: string;
  cryptoSymbols?: string;
  walletAddress?: string;
  balanceAlgo?: number;
  runUrl?: string;
  status?: "draft" | "live";
  executionState?: "idle" | "running" | "done" | "error";
  executionNote?: string;
  executionMessage?: string;
  executionOutput?: string;
  onFundWallet?: (nodeId: string) => void;
  onCopyWallet?: (nodeId: string) => void;
  onTriggerTestChange?: (nodeId: string, value: string) => void;
  onApiKeyChange?: (nodeId: string, value: string) => void;
}

export interface PipelineEdgeData extends Record<string, unknown> {
  wireType: WireKind;
  label?: string;
}

export type BuilderNode = Node<PipelineNodeData>;
export type BuilderEdge = Edge<PipelineEdgeData>;

export interface DeployRequest {
  name: string;
  network: "algorand-testnet";
  nodes: Array<{
    id: string;
    type: NodeKind;
    position: { x: number; y: number };
    data: Omit<
      PipelineNodeData,
      | "kind"
      | "walletAddress"
      | "balanceAlgo"
      | "runUrl"
      | "status"
      | "executionState"
      | "executionNote"
      | "executionMessage"
      | "executionOutput"
      | "onFundWallet"
      | "onCopyWallet"
      | "onTriggerTestChange"
      | "onApiKeyChange"
    >;
  }>;
  edges: Array<{
    id: string;
    source: string;
    target: string;
    sourceHandle?: string;
    targetHandle?: string;
    wireType: WireKind;
    label?: string;
  }>;
}

export interface LogEntry {
  timestamp: string;
  level: LogLevel;
  message: string;
  nodeId?: string;
  eventType?: "start" | "progress" | "output" | "done" | "error";
  output?: string;
  txId?: string;
  details?: Record<string, string | number | boolean | null>;
}

export interface DeployedNodeState {
  id: string;
  type: NodeKind;
  walletAddress?: string;
  balanceAlgo?: number;
  explorerUrl?: string;
}

export interface DeployResponse {
  pipelineId: string;
  status: "live";
  endpoint: string;
  priceAlgo: number;
  network: string;
  paymentWallet?: string;
  loraUrl?: string;
  nodes: DeployedNodeState[];
  logs: LogEntry[];
}

export interface PipelineSummary {
  pipelineId: string;
  name: string;
  endpoint: string;
  network: string;
  priceAlgo: number;
  paymentWallet?: string;
  nodeCount: number;
  wireCount: number;
  runCount: number;
}

export interface PipelineDetail {
  pipelineId: string;
  definition: DeployRequest;
  endpoint: string;
  priceAlgo: number;
  network: string;
  paymentWallet?: string;
  loraUrl?: string;
  nodes: DeployedNodeState[];
}

export interface BalanceRecord {
  nodeId: string;
  address: string;
  balanceAlgo: number;
  explorerUrl?: string;
}

export interface BalanceResponse {
  pipelineId: string;
  balances: BalanceRecord[];
}

export interface FundIntentResponse {
  pipelineId: string;
  nodeId: string;
  address: string;
  faucetUrl: string;
  loraUrl?: string;
  qrValue: string;
}

export interface RunResponse {
  runId: string;
  pipelineId: string;
  result: string;
  settlementMode: "demo" | "payment_response";
  logs: LogEntry[];
}

export interface PaymentPreflightResponse {
  status: number;
  paymentRequired: boolean;
  facilitator?: string;
  body: {
    pipelineId: string;
    amountAlgo: number;
    wallet: string;
    network: string;
    facilitator: string;
    endpoint: string;
    message: string;
  };
}
